xmlua/c14n.lua

```diff
@@ -0,0 +1,237 @@
+    end,
+    [ffi.C.XML_DTD_NODE] = function(document, raw_node)
+      error("XML_DTD_NODE not implemented")               -- TODO: implement
+    end,
+    [ffi.C.XML_ELEMENT_DECL] = function(document, raw_node)
+      return ElementDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_ATTRIBUTE_DECL] = function(document, raw_node)
+      return AttributeDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_ENTITY_DECL] = function(document, raw_node)
+      return EntityDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_NAMESPACE_DECL] = function(document, raw_node)
+      return NamespaceDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_XINCLUDE_START] = function(document, raw_node)
+      error("XML_XINCLUDE_START not implemented")         -- TODO: implement
+    end,
+    [ffi.C.XML_XINCLUDE_END] = function(document, raw_node)
+      error("XML_XINCLUDE_END not implemented")           -- TODO: implement
+    end,
+    [ffi.C.XML_DOCB_DOCUMENT_NODE] = function(document, raw_node)
+      error("XML_DOCB_DOCUMENT_NODE not implemented")     -- TODO: implement
+    end,
+  }, {
+    __index = function(self, key)
+      error("Unknown node type: " .. tostring(key))
+    end
+  })
+
+  function wrap_raw_node(document, raw_node)
+    if raw_node == ffi.NULL then
+      return nil
+    end
+    return type_generators[tonumber(raw_node.type)](document, raw_node)
+  end
+end
+
+
+
+--- Canonicalise an xmlDocument or set of elements.
+-- @param self xmlDoc from which to canonicalize elements
+-- @param select array of nodes to include, or function to determine if a node should be included in the canonicalized output.
+--        Signature: `boolean = function(node, parent)`
+-- @param mode any of C14N_1_0, C14N_EXCLUSIVE_1_0 (default), C14N_1_1
+-- @param inclusive_ns_prefixes array, or space-separated string, of namespace prefixes to include
+-- @param with_comments if truthy, comments will be included (default: false)
+-- @return string containing canonicalized xml
+function C14n:canonicalize(select, mode, inclusive_ns_prefixes, with_comments)
+  if mode == nil then
+    mode = DEFAULT_MODE
+  end
+  with_comments = with_comments and 1 or 0 -- default = not including comments
+
+  local callback = type(select) == "function" and select or nil
+  local nodes = type(select) == "table" and select or nil
+  assert(callback or nodes, "select must be a function or an array of nodes")
+
+  mode = assert(C14N_MODES_LOOKUP[mode], "mode must be a valid C14N mode constant")
+
+  local prefixes = getNamespacePrefixArray(inclusive_ns_prefixes)
+  local buffer = libxml2.xmlBufferCreate()
+  local output_buffer = libxml2.xmlOutputBufferCreate(buffer)
+
+  local success
+  if callback then
+    -- wrap the callback to pass wrapped objects, and return 1 or 0
+    local cbwrapper = function(_, nodePtr, parentPtr)
+      return callback(wrap_raw_node(self, nodePtr), wrap_raw_node(self, parentPtr)) and 1 or 0
```

Comment:
Could you use this style for easy to read?

```suggestion
      if callback(wrap_raw_node(self, nodePtr), wrap_raw_node(self, parentPtr)) then
        return 1
      else
        return 0
      end
```

Reply:
done